iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relocatable_ptr.inl

```diff
@@ -0,0 +1,164 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_HOOFS_RELOCATABLE_POINTER_RELOCATABLE_PTR_INL
+#define IOX_HOOFS_RELOCATABLE_POINTER_RELOCATABLE_PTR_INL
+
+#include "iceoryx_hoofs/internal/relocatable_pointer/relocatable_ptr.hpp"
+
+namespace iox
+{
+namespace rp
+{
+template <typename T>
+relocatable_ptr<T>::relocatable_ptr(T* ptr)
+{
+    m_offset = to_offset(ptr);
+}
+
+template <typename T>
+relocatable_ptr<T>::relocatable_ptr(const relocatable_ptr& other)
+{
+    m_offset = to_offset(other.get());
+}
+
+template <typename T>
+relocatable_ptr<T>::relocatable_ptr(relocatable_ptr&& other)
+{
+    m_offset = to_offset(other.get());
+    other.m_offset = NULL_POINTER_OFFSET;
+}
+
+template <typename T>
+relocatable_ptr<T>& relocatable_ptr<T>::operator=(const relocatable_ptr& rhs)
+{
+    if (this != &rhs)
+    {
+        m_offset = to_offset(rhs.get());
+    }
+    return *this;
+}
+
+template <typename T>
+relocatable_ptr<T>& relocatable_ptr<T>::operator=(relocatable_ptr&& rhs)
+{
+    if (this != &rhs)
+    {
+        m_offset = to_offset(rhs.get());
+        rhs.m_offset = NULL_POINTER_OFFSET;
+    }
+    return *this;
+}
+
+template <typename T>
+T* relocatable_ptr<T>::get()
+{
+    return from_offset(m_offset);
+}
+
+template <typename T>
+const T* relocatable_ptr<T>::get() const
+{
+    return from_offset(m_offset);
+}
+
+template <typename T>
+template <typename S>
+S& relocatable_ptr<T>::operator*()
+{
+    // not actually evaluated in the error case (compiler fails earlier since S = void leads to void&)
+    static_assert(!std::is_same<S, void>::value, "relocatable_ptr<void> does not support operator*");
+    return *get();
+}
+
+template <typename T>
+template <typename S>
+const S& relocatable_ptr<T>::operator*() const
+{
+    // not actually evaluated in the error case (compiler fails earlier since S = void leads to void&)
+    static_assert(!std::is_same<S, void>::value, "relocatable_ptr<void> does not support operator* const");
+    return *get();
+}
+
+template <typename T>
+T* relocatable_ptr<T>::operator->()
+{
+    return get();
+}
+
+template <typename T>
+const T* relocatable_ptr<T>::operator->() const
+{
+    return get();
+}
+
+template <typename T>
+relocatable_ptr<T>::operator T*()
+{
+    return get();
+}
+
+template <typename T>
+relocatable_ptr<T>::operator const T*() const
+{
+    return get();
+}
+
+template <typename T>
+typename relocatable_ptr<T>::offset_t relocatable_ptr<T>::self() const
+{
+    return reinterpret_cast<offset_t>(this);
+}
+
+template <typename T>
+typename relocatable_ptr<T>::offset_t relocatable_ptr<T>::to_offset(const void* ptr) const
+{
+    if (ptr == nullptr)
+    {
+        return NULL_POINTER_OFFSET;
+    }
+    auto p = reinterpret_cast<offset_t>(ptr);
+    return p - self();
```

Comment:
You could also write a fancy one liner here:
```cpp
return ( ptr ) ? reinterpret_cast<offset_t>(ptr) - self() : nullptr; 
```

Reply:
Sure, but I am fairly certain the compiler will create the same code regardless and I prefer not to crowded lines myself when it is not necessary.